File: getting-started-mocks/07-mock-the-wire/node/notifier/carrier.js
import { CarrierUnavailableError, TimeoutError } from './errors.js'

export class CarrierClient {
  #baseUrl
  #apiKey

  constructor(baseUrl, apiKey) {
    this.#baseUrl = baseUrl.replace(/\/+$/, '')
    this.#apiKey = apiKey
  }

  async lookup(trackingNumber) {
    let response
    try {
      response = await fetch(`${this.#baseUrl}/shipments/${encodeURIComponent(trackingNumber)}`, {
        headers: {
          'X-API-Key': this.#apiKey,
          Accept: 'application/json',
        },
        signal: AbortSignal.timeout(2000),
      })
    } catch (failure) {
      if (failure.name === 'TimeoutError' || failure.name === 'AbortError') {
        throw new TimeoutError(`carrier did not answer in time: ${failure.message}`)
      }
      throw failure
    }

    if (!response.ok) {
      throw new CarrierUnavailableError(`carrier returned ${response.status}`)
    }
    const shipment = await response.json()
    return shipment.status
  }
}

File: getting-started-mocks/07-mock-the-wire/python/notifier/carrier.py
"""The real carrier client, pointed at a test-controlled server by wire tests."""

import json
from urllib.error import HTTPError
from urllib.parse import quote
from urllib.request import Request, urlopen

from notifier.errors import CarrierUnavailableError

TIMEOUT_SECONDS = 2


class CarrierClient:
    def __init__(self, base_url: str, api_key: str) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key

    def lookup(self, tracking_number: str) -> str:
        request = Request(
            f"{self._base_url}/shipments/{quote(tracking_number, safe='')}",
            headers={"X-API-Key": self._api_key, "Accept": "application/json"},
        )
        try:
            with urlopen(request, timeout=TIMEOUT_SECONDS) as response:
                shipment = json.load(response)
        except HTTPError as failure:
            raise CarrierUnavailableError(f"carrier returned {failure.code}") from failure
        return shipment["status"]


def http_shipment_status(base_url: str, api_key: str = ""):
    """Compatibility helper that exposes the client's lookup seam."""
    return CarrierClient(base_url, api_key).lookup
